contracts/test/calldata.spec.ts

```diff
@@ -0,0 +1,92 @@
+// SPDX-License-Identifier: Apache-2.0
+
+import { expect } from 'chai';
+import { ethers } from 'hardhat';
+import { TestCalldataEncryption } from '../typechain-types/contracts/tests';
+import {
+  boxKeyPairFromSecretKey,
+  crypto_box_SECRETKEYBYTES,
+  isCalldataEnveloped,
+  X25519DeoxysII,
+} from '@oasisprotocol/sapphire-paratime';
+import { hexlify, parseUnits, randomBytes } from 'ethers';
+import { randomInt } from 'crypto';
+
+describe('CalldataEncryption', () => {
+  let contract: TestCalldataEncryption;
+
+  before(async () => {
+    const factory = await ethers.getContractFactory('TestCalldataEncryption');
+    contract = await factory.deploy();
+    await contract.waitForDeployment();
+  });
+
+  // Ensures that the JS library provides the same results as Solidity
+  it('testEncryptCallData', async () => {
+    for (let i = 1; i < 1024; i += 1 + i / 5) {
+      const peerKeypair = boxKeyPairFromSecretKey(
+        randomBytes(crypto_box_SECRETKEYBYTES),
+      );
+      const myKeypair = boxKeyPairFromSecretKey(
+        randomBytes(crypto_box_SECRETKEYBYTES),
+      );
+      const calldata = randomBytes(i);
+      const epoch = randomInt(1 << 32);
+      const nonce = randomBytes(15);
+      const cipher = X25519DeoxysII.fromSecretKey(
+        myKeypair.secretKey,
+        peerKeypair.publicKey,
+        epoch,
+      );
+      const encryptedCall = cipher.encryptCall(calldata, nonce);
+      const result = await contract.testEncryptCallData(
+        calldata,
+        myKeypair.publicKey,
+        myKeypair.secretKey,
+        nonce,
+        epoch,
+        peerKeypair.publicKey,
+      );
+      expect(result).eq(hexlify(encryptedCall));
+    }
+  });
+
+  it('roundtrip encryption', async () => {
+    // Tests must be submitted from an account which has a balance
+    // But can't get access to the signer private key from here
+    // So, assume the 0xf39F address is being used to run tests
+    const myAddr = '0xf39Fd6e51aad88F6F4ce6aB8827279cffFb92266';
+    const myKey =
+      '0xac0974bec39a17e36ba4a6b4d238ff944bacb478cbed5efcae784d7bf4f2ff80';
+    const [signerAddr] = await ethers.getSigners();
```

Comment:
Initialisation could be moved in the `before` block, for better structure and if more tests are added in future that need this values too.